### PR TITLE
fix: osnap improve how osnap form validates trarnsactions

### PR DESCRIPTION
**Pull request by gsteenkamp89**
## motivation

disables publish until ALL transactions in the builder are valid

### CODE REVIEW

File: src/plugins/oSnap/components/Input/Amount.vue [hunk 1]
```diff
@@ -16,24 +16,31 @@ const dirty = ref(false);
 
 const format = (amount: string) => {
   try {
-    return parseUnits(amount, props.decimals).toString();
+    // empty string throws
+    const parsed = parseUnits(amount, props.decimals).toString();
+    return parsed;
   } catch (error) {
-    return undefined;
+    return parseUnits('0', props.decimals).toString();
```
Comment (daywiss) [inline src/plugins/oSnap/components/Input/Amount.vue:23]: do you actually need to call parse units on 0? cant you just put '0'?

Comment (daywiss) [reply src/plugins/oSnap/components/Input/Amount.vue:23]: also im not sure if we should really be resetting input box to 0 if theres an error

File: src/plugins/oSnap/components/TransactionBuilder/TransferFunds.vue [hunk 2]
```diff
@@ -55,10 +54,10 @@ function updateTransaction() {
...
     });
+    emit('updateTransaction', transaction);
     const isTransactionValid = validateTransaction(transaction);
-    if (isTransactionValid) {
-      emit('updateTransaction', transaction);
-      return;
+    if (!isTransactionValid) {
+      console.warn('invalid transaction');
     }
   } catch (error) {
```
Comment (daywiss) [inline src/plugins/oSnap/components/TransactionBuilder/TransferFunds.vue:62]: what happens here? if transaction errors the parent never sees this is invalid

Comment (gsteenkamp89) [reply src/plugins/oSnap/components/TransactionBuilder/TransferFunds.vue:62]: can remove conditional entirely

File: src/plugins/oSnap/components/TransactionBuilder/ContractInteraction.vue [hunk 3]
```diff
@@ -24,7 +24,7 @@ const emit = defineEmits<{
 
 const to = ref(props.transaction.to ?? '');
 const isToValid = computed(() => {
-  return to.value === '' || isAddress(to.value);
+  return to.value !== '' && isAddress(to.value);
```
Comment (gsteenkamp89) [inline src/plugins/oSnap/components/TransactionBuilder/ContractInteraction.vue:27]: oh i need to revert this change, since `value` should be allowed to be `0`

File: src/plugins/oSnap/components/TransactionBuilder/ContractInteraction.vue [hunk 4]
```diff
@@ -40,22 +40,27 @@ const selectedMethod = computed(
...
-
-    if (validateTransaction(transaction)) {
-      emit('updateTransaction', transaction);
-      return;
-    }
-  } catch (error) {
-    console.warn('ContractInteraction - Invalid Transaction:',error);
+    emit('updateTransaction', transaction);
+  } catch {
+    props.setTransactionAsInvalid();
```
Comment (gsteenkamp89) [inline src/plugins/oSnap/components/TransactionBuilder/ContractInteraction.vue:63]: this is essentially what I'm doing in each transaction type.
If validation fails, we throw, then in the catch block we keep the data the same upstream, but we add an invalid flag.
the form doesn't lose it's state. so as soon as the user inputs data that doesn't cause the `create<TransactionType>` to throw, we can update state upstream with the formatted Tx and set it to be valid
